service/storage_proxy.cc

```diff
@@ -723,6 +723,56 @@ class storage_proxy::remote {
         auto cmd = make_lw_shared<query::read_command>(std::move(cmd1));
         auto src_ip = src_addr.addr;
         auto timeout = t ? *t : db::no_timeout;
+
+        // Verify whether read_command is provided in legacy or native reversed format by comparing
+        // schema version with table schema version. There are three options:
+        // 1. Versions are the same -> legacy format
+        // 2. schema version is equal to reversed table schema -> native format
+        // 3. table schema is outdated
+        static auto is_read_command_in_reversed_legacy_format = [](const query::read_command& cmd, schema_ptr table_schema) -> std::optional<bool> {
+            if (cmd.schema_version == table_schema->version()) {
+                return true;
+            }
+
+            if (cmd.schema_version == reversed(table_schema->version())) {
+                return false;
+            }
+
+            return std::nullopt;
+        };
+
+        // Detect whether a transformation from legacy reverse format into native reverse
+        // format is necessary before executing the read_command. That happens when the
+        // native_reverse_queries feature is turned off. A custer has mixed nodes.
+        auto f_format_reverse_required = co_await coroutine::as_future([&]() -> future<bool> {
```

Comment:
This is adding an allocation in a very hot path. Better to fold the inner coroutine into the outer. Or better, handle the common case (fully updated cluster) inline and throw the edge case into a helper (which can be slower).